src/librbd/image/CreateRequest.cc

```diff
@@ -0,0 +1,795 @@
+
+  m_mirror_image_internal.state = cls::rbd::MIRROR_IMAGE_STATE_ENABLED;
+  if (m_non_primary_global_image_id.empty()) {
+    uuid_d uuid_gen;
+    uuid_gen.generate_random();
+    m_mirror_image_internal.global_image_id = uuid_gen.to_string();
+  } else {
+    m_mirror_image_internal.global_image_id = m_non_primary_global_image_id;
+  }
+
+  librados::ObjectWriteOperation op;
+  cls_client::mirror_image_set(&op, m_imageid, m_mirror_image_internal);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_mirror_image_enable>(this);
+  int r = m_ioctx.aio_operate(RBD_MIRRORING, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_mirror_image_enable(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "cannot enable mirroring: " << cpp_strerror(*result) << dendl;
+
+    m_r_saved = *result;
+    journal_remove();
+    return nullptr;
+  }
+
+  send_watcher_notification();
+  return nullptr;
+}
+
+// TODO: make this *really* async
+template<typename I>
+void CreateRequest<I>::send_watcher_notification() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  Context *ctx = new FunctionContext([this](int r) {
+      r = MirroringWatcher<>::notify_image_updated(
+        m_ioctx, cls::rbd::MIRROR_IMAGE_STATE_ENABLED,
+        m_imageid, m_mirror_image_internal.global_image_id);
+      handle_watcher_notify(r);
+    });
+
+    m_op_work_queue->queue(ctx, 0);
+}
+
+template<typename I>
+void CreateRequest<I>::handle_watcher_notify(int r) {
+  ldout(m_cct, 20) << __func__ << ": r=" << r << dendl;
+
+  if (r < 0) {
+    // non fatal error -- watchers would cope up upon noticing missing
+    // updates. just log and move on...
+    ldout(m_cct, 10) << "failed to send update notification: " << cpp_strerror(r)
+                     << dendl;
+  }
+
+  ldout(m_cct, 20) << "image mirroring is enabled: global_id=" <<
+    m_mirror_image_internal.global_image_id << dendl;
+  complete(0);
+}
+
+template<typename I>
+void CreateRequest<I>::complete(int r) {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  m_r_saved = r;
+
+  using klass = CreateRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::finish>(this);
+
+  shut_down_and_cleanup_journaler(ctx);
+}
+
+template<typename I>
+Context *CreateRequest<I>::finish(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "failed to shut down journaler: " << cpp_strerror(*result) << dendl;
+  }
+
+  *result = m_r_saved;
+  if (*result == 0) {
+    ldout(m_cct, 20) << "done." << dendl;
+  }
+  return m_on_finish;
+}
+
+// helper
+template<typename I>
+void CreateRequest<I>::shut_down_and_cleanup_journaler(Context *ctx) {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
```

Comment:
This can be removed once the lifetime of journal::Journaler is owned by the journal::Create/RemoveRequest state machines.